docs/source/model_library.rst

```diff
@@ -0,0 +1,550 @@
+
+.. code-block:: pycon
+
+   >>> library = ModelLibrary("my_asn.json")
+
+.. NOTE::
+   For an association that contains a ``group_id`` for each member,
+   creating a library will not read any models.
+
+.. _library_borrowing_and_shelving:
+
+Borrowing and shelving models
+-----------------------------
+
+Interacting with an `~stpipe.library.AbstractModelLibrary` involves "borrowing" and "shelving"
+models, both of which must occur during a ``with`` statement (while the library
+is "open"):
+
+.. code-block:: pycon
+
+   >>> with library:
+   ...    model = library.borrow(0)
+   ...    # do stuff with the model...
+   ...    library.shelve(model)
+
+Iteration is also supported (but don't forget to return your models!).
+
+.. code-block:: pycon
+
+   >>> with library:
+   ...    for model in library:
+   ...        # do stuff with the model...
+   ...        library.shelve(model)
+
+
+.. _library_on_disk:
+
+On Disk Mode
+------------
+
+For large associations (like those larger than memory) it is important
+that the library avoid reading all models at once. The borrow/shelve API
+above maps closely to the loading/saving of input (or tempoary) files
+containing the models.
+
+.. code-block:: pycon
+
+   >>> library = ModelLibrary("my_big_asn.json", on_disk=True)
+   >>> with library:
+   ...     model = library.borrow(0)  # the input file for model 0 is loaded
+   ...     library.shelve(model)  # a temporary file for model 0 is written
+
+.. NOTE::
+   In the above example, a temporary file was created for model 0. At no
+   point will the library overwrite the input file.
+
+If model is not modified during the time it's borrowed (for example if the
+``model.dq`` array was read, but not modified). It is helpful to tell the
+library that the model was not modified.
+
+.. code-block:: pycon
+
+   >>> with library:
+   ...     model = library.borrow(0)  # the input file for model 0 is loaded
+   ...     # do some read-only stuff with the model
+   ...     library.shelve(model, modify=False)  # No temporary file will be written
+
+This can dramatically reduce the number of times a file is written saving
+on both disk space and the time required to write.
+
+
+.. _library_map_function:
+
+Map function
+------------
+
+Let's say you want to get the ``meta.filename`` attribute for all models
+in a library. The above "open", "borrow", "shelve" pattern can be quite
+verbose. Instead, the helper method `~stpipe.library.AbstractModelLibrary.map_function`
+can be used to generate an iterator that returns the result of a function
+applied to each model in the library:
+
+.. code-block:: pycon
+
+   >>> def get_model_name(model, index):
+   ...     return model.meta.filename
+   >>>
+   >>> filenames = list(library.map_function(get_model_name))
+
+.. NOTE::
+   `~stpipe.library.AbstractModelLibrary.map_function` does not require an open library
+   and will handle opening, borrowing, shelving and closing for you.
+
```

Comment:
might be good to specify here that the arguments to map_function MUST be `model, index`.  And random question too, does it need to take exactly those arguments, or can you have any number of arguments as long as those are the first two?

Reply:
Is the docstring for the function sufficient? If not I'd rather expand on that (since it's what will be seen in an interactive session).

At the moment the only supported function signature is `model, index`. Would you open a separate issue to discuss changes to this? I think it's worth considering changing this as we get more experience with `ModelLibrary` including:
- making `index` optional
- adding optional `args` and `kwargs` that get passed through `map_function`

Reply:
yes, can do.  I haven't played with `map_function` myself yet, but I will try to replace some of my explicit `borrow` and `shelve` loops with it in the JWST PR.  Then it will be easier to point to specific examples where this functionality might be useful